### PR TITLE
Fix #3245,#2193,#2941,#5143  - Timeline: Migrate to vis.js

### LINKED ISSUE #3245: Timeline: Migrate to vis.js
http://almende.github.io/chap-links-library/

> CHAP Links Library now has a successor: vis.js. A couple of fundamental limitations in the current architecture of CHAP Links Library have been reached. Therefore, a new initiative has been started, aiming at more flexible data manipulation, better performance, and expandable visualization tools.

http://visjs.org/

### PR DESCRIPTION
**Pull request by jasonex7**
Fixes #3245 

- Using vis.js 5.1. It was recommended to upgrade to v6 which it should be out in the next few days. It was told that it will be 95% compatible to current 5.1 release so I have only to replace 0-timeline.min.js. (https://github.com/visjs/vis-timeline)
- TimelineEvent, TimelineGroup, TimelineModel, TimelineModificationEvent, TimelineSelectEvent now use generics.
- TimelineEvent has now 3 more properties to fine tunning editable: editableTime, editableGroup and editableRemove.
- TimelineGroup has now styleClass property similar to the one in TimelineEvent.
- Properties deprecated in favor of others: axisOnTop, groupsChangeable, groupsWidth, groupMinHeight, snapEvents, timeChangeable.
- Properties deprecated with NO replacement: dragAreaWith (use css classes instead), unselectable, groupsOnRight, animate, animateZoom, showButtonNew, showNavigation.
- New properties: maxHeight, orientationAxis, orientationItem, editableAdd, editableTime, editableGroup, editableRemove, editableOverrideItems, eventHorizontalMargin, eventVerticalMargin, groupStyle, snap, clickToUse.
- New extender property to specify a javascript function to extend configuration of vis.js timeline component similar to extender property in <p:chart> component.

### CODE REVIEW

**Comment by jasonex7** [comment]
I recommend create a new facet called maybe "navigator" or "menu" so we can add custom buttons to add functionality that was deprecated by properties `showButtonNew` and `showNavigation` similiar to this [example](https://visjs.github.io/vis-timeline/examples/timeline/interaction/navigationMenu.html). What do you think @melloware?

**Comment by christophs78** [comment]
Happy to see this PR. Just started yesterday on moving timeline to java.time.LocalDateTime and noted the old library. I'll have to redo some work after this PR is merged, but I think it is worth the work.

File: src/main/java/org/primefaces/component/timeline/TimelineBase.java [hunk 1]
```diff
@@ -172,21 +182,29 @@ public void setBrowserTimeZone(Object browserTimeZone) {
     }
 
     public String getHeight() {
-        return (String) getStateHelper().eval(PropertyKeys.height, "auto");
+        return (String) getStateHelper().eval(PropertyKeys.height, null);
```
Comment (melloware) [inline src/main/java/org/primefaces/component/timeline/TimelineBase.java:190]: Is "auto" no longer a valid option in the new timeline?  What are the options for height?

File: src/main/java/org/primefaces/component/timeline/DefaultTimelineUpdater.java [hunk 2]
```diff
@@ -119,44 +121,72 @@ public void beforePhase(PhaseEvent event) {
...
+                    for (int i = 0; i < groups.size(); i++) {
+                        TimelineGroup<?> group = groups.get(i);
+                        if (group.getId() != null && group.getId().equals(data.getEvent().getGroup())) {
+                            orderGroup = i;
+                            foundGroup = group;
+                            break;
+                        }
+                    }
+                    if (foundGroup != null) {
+                        //If groups was not setted in model then order by content.
```
Comment (melloware) [inline src/main/java/org/primefaces/component/timeline/DefaultTimelineUpdater.java]: "setted" should be "set"

File: src/main/java/org/primefaces/component/timeline/TimelineBase.java [hunk 3]
```diff
@@ -50,13 +54,18 @@
...
-        dragAreaWidth,
+        orientationAxis,
+        orientationItem,
         editable,
+        editableAdd,
+        editableRemove,
+        editableGroup,
+        editableTime,
+        editableOverrideItems,
         selectable,
```
Comment (melloware) [inline src/main/java/org/primefaces/component/timeline/TimelineBase.java:68]: Need to update /docs/7.1/components/timeline to add and remove all these properties.

File: src/main/java/org/primefaces/component/timeline/TimelineRenderer.java [hunk 4]
```diff
@@ -85,42 +86,97 @@ protected void encodeScript(FacesContext context, Timeline timeline) throws IOEx
...
             groupsContent = new HashMap<>();
         }
 
-        List<TimelineEvent> events = model.getEvents();
+        UIComponent groupFacet = timeline.getFacet("group");
+        int groupsSize = groups.size();
+        if (groupsSize > 0) {
+            writer.write(",groups:[");
+            for (int i = 0; i < groupsSize; i++) {
+                //If groups was not setted in model then order by content.
```
Comment (melloware) [inline src/main/java/org/primefaces/component/timeline/TimelineRenderer.java]: "setted" = "set"

File: src/main/java/org/primefaces/component/timeline/TimelineRenderer.java [hunk 5]
```diff
@@ -141,174 +197,163 @@ protected void encodeScript(FacesContext context, Timeline timeline) throws IOEx
...
         }
 
-        if (event.isEditable() != null) {
-            fsw.write(",\"editable\":" + event.isEditable());
-        }
-        else {
-            fsw.write(",\"editable\":null");
+        if (event.isEditableTime() != null || event.isEditableGroup() != null || event.isEditableRemove() != null) {
+            fsw.write(", editable: {");
+            boolean wasSetted = false;
```
Comment (melloware) [inline src/main/java/org/primefaces/component/timeline/TimelineRenderer.java]: wasSetted = isSet

File: src/main/java/org/primefaces/component/timeline/TimelineRenderer.java [hunk 6]
```diff
@@ -141,174 +197,163 @@ protected void encodeScript(FacesContext context, Timeline timeline) throws IOEx
...
-            // there is a list of groups ==> find the group to the event by the group id
-            TimelineGroup foundGroup = null;
-            int groupOrder = 0;
-            for (TimelineGroup group : groups) {
+        // there is a list of groups ==> find the group to the event by the group id
+        TimelineGroup<?> foundGroup = null;
+        if (event.getGroup() != null) {
+            for (TimelineGroup<?> group : groups) {
                 if (group.getId() != null && group.getId().equals(event.getGroup())) {
                     foundGroup = group;
```
Comment (melloware) [inline src/main/java/org/primefaces/component/timeline/TimelineRenderer.java:376]: if (Objects.equals(group.getId(), event.getGroup)) {

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by melloware** [comment]
> I recommend create a new facet called maybe "navigator" or "menu" so we can add custom buttons to add functionality that was deprecated by properties `showButtonNew` and `showNavigation` similiar to this [example](https://visjs.github.io/vis-timeline/examples/timeline/interaction/navigationMenu.html). What do you think @melloware?

I like that idea.

File: src/main/java/org/primefaces/component/timeline/TimelineBase.java (hunk 1 above)
Comment (jasonex7) [reply src/main/java/org/primefaces/component/timeline/TimelineBase.java:190]: No, "auto" is no longer a valid options, that was something that give me an error on tests. If the property is omitted it behaves like auto (see next image of description of height property). 

image

File: src/main/java/org/primefaces/component/timeline/TimelineBase.java (hunk 3 above)
Comment (jasonex7) [reply src/main/java/org/primefaces/component/timeline/TimelineBase.java:68]: By now I don't remove from taglib to mantain compatibility with existing xhtml code, so I only removed these from `TimelineBase`. I documented on taglib that the properties was deprecated.

File: src/main/java/org/primefaces/component/timeline/DefaultTimelineUpdater.java (hunk 2 above)
Comment (jasonex7) [reply src/main/java/org/primefaces/component/timeline/DefaultTimelineUpdater.java]: Sorry my bad English on that 😅

File: src/main/java/org/primefaces/component/timeline/TimelineBase.java (hunk 3 above)
Comment (melloware) [reply src/main/java/org/primefaces/component/timeline/TimelineBase.java:68]: That is good but the actual Markdown doc need to be update in the /docs/7.1/ folder.

**Comment by melloware** [comment]
Adding link to this ticket which we should also do with this upgrade: https://github.com/primefaces/primefaces/issues/2941

File: src/main/java/org/primefaces/component/timeline/TimelineBase.java (hunk 3 above)
Comment (jasonex7) [reply src/main/java/org/primefaces/component/timeline/TimelineBase.java:68]: > That is good but the actual Markdown doc need to be update in the /docs/7.1/ folder.

That's a new world for me haha. Check this new commit.

**Comment by melloware** [comment]
I just assigned https://github.com/primefaces/primefaces/issues/2193 to make the TimeLine RTL aware! Should be easy addition for you. 😄

**Comment by jasonex7** [comment]
> I just assigned #2193 to make the TimeLine RTL aware! Should be easy addition for you. 😄

You almost read my mind haha. I'm finishing menu facet supporting a new dir (ltr or rtl) property.

**Comment by jasonex7** [comment]
> Adding link to this ticket which we should also do with this upgrade: #2941

In vis.js documentation says that localization depends on the localization of `moment.js` library. So the user must load a localized version of moment.js (like [this example](https://visjs.github.io/vis-timeline/examples/timeline/other/localizationCustom.html)) and use `locale` property on `<p:timeline>` or set that locale on the jsf view.

> ### Localization
> Timeline can be localized. For localization, Timeline depends largely on the localization of moment.js. Locales are not included in vis.js by default. To enable localization, moment.js must be loaded with locales. Moment.js offers a bundle named "moment-with-locales.min.js" for this and there are various alternative ways to load locales.

**Comment by melloware** [comment]
> In vis.js documentation says that localization depends on the localization of `moment.js` library. So the user must load a localized version of moment.js (like [this example](https://visjs.github.io/vis-timeline/examples/timeline/other/localizationCustom.html)) and use `locale` property on `<p:timeline>` or set that locale on the jsf view.

Maybe we should put this information on the /docs/7.1/ page for users who will definitely ask how to localize it.

**Review by melloware — APPROVED** [review]

**Comment by jasonex7** [comment]
> Maybe we should put this information on the /docs/7.1/ page for users who will definitely ask how to localize it.

OK, I will work on putting that on docs. I also open the issue https://github.com/visjs/vis-timeline/issues/78 on vis.js to ask for localization of delete button, so maybe in a next release (hope in v6 😄) we should update that on docs.

**Comment by melloware** [comment]
@jasonex7 this was just fixed this morning: https://github.com/primefaces/primefaces/issues/5118

Can you review and make the same fix in your branch?

**Comment by jasonex7** [comment]
> @jasonex7 this was just fixed this morning: #5118
> 
> Can you review and make the same fix in your branch?

I think that is not an issue, but a problem of lack of documentation. The documentation of timeline component when it was on PrimeFaces Extensions was more explained than when it was moved to PrimeFaces. See my comment https://github.com/primefaces/primefaces/issues/5118#issuecomment-528891817

**Comment by melloware** [comment]
I see that. Thanks for investigating.

**Review by melloware — APPROVED** [review]
This is a huge improvement and will prevent future bugs by introducing the id value. Great job!

**Comment by jasonex7** [comment]
I've just merged PR #5157 that fixes #5149 in this PR.

**Review by melloware — APPROVED** [review]

**Comment by tandraschko** [comment]
V6 already released?
Does the migration break existing apps?

**Comment by melloware** [comment]
@tandraschko we are still awaiting the release of v6 which fixes an IE11 issue.  @jasonex7 has done eveyrthing he can to make it backwards compatible and marked which methods are deprecated in both the taglib and code.  Similar to how we did Schedule upgrade in 7.0 I think we will just have to put these notes in the Migration Guide for which props are deprecated or replaced etc.

**Comment by tandraschko** [comment]
any news?

**Comment by melloware** [comment]
I am monitoring their releases and they have not released 6.0 yet.  They have fixed a lot of issues and @jasonex7 has submitted quite a few PR's that have been accepted.  It should be soon though.

**Comment by jasonex7** [comment]
As a comment, we're using the current vis-timeline build in a production app and we have no problems and works perfectly in IE11 thanks to visjs/vis-timeline#72 got fixed. So maybe, a custom build could be included with this and others PR that have been submitted to visjs/vis-timeline repo, what do you think?

**Comment by melloware** [comment]
@jasonex7 I am OK with that so it will let us migrate this PR and allow you/us to work on the showcase examples migration from PF Extensions.  then before we cut 7.1 or whenever its released we can just push the final version of vis timeline.  So I approve.

**Comment by tandraschko** [comment]
thats ok for me now.
we should then upgrade the scripts ASAP when its released.

Can you update the PR with a snapshot build from vis?

**Comment by jasonex7** [comment]
> Can you update the PR with a snapshot build from vis?

Done.

**Comment by melloware** [comment]
Thanks @jasonex7 its now merged!

**Comment by tandraschko** [comment]
hey guys (@jasonex7 @melloware)
if you open the showcase (http://localhost:8080/showcase/ui/data/timeline/linked.xhtml)
and zoom or drag something, a endless loop of AJAX requests will be fired
can you please fix that?

**Comment by tandraschko** [comment]
And please also change the migrationguide

**Comment by melloware** [comment]
Migration Guide has been updated.

**Comment by jasonex7** [comment]
> if you open the showcase (http://localhost:8080/showcase/ui/data/timeline/linked.xhtml)
> and zoom or drag something, a endless loop of AJAX requests will be fired
> can you please fix that?

I've investigated and that error appears also with the previous timeline version (chap-links-library). There are plenty of ajax requests, but finally they eventually stopped. I will create another issue to work on that.

**Pull request merged**